### PR TITLE
Fix tag commit message

### LINKED ISSUE #21687: Incorrect latest commit message in UI when viewing tag
### Description

The latest commit is incorrectly rendered when viewing a tag, the first line (commit title) is absent. It's okay when viewing a branch.

Not OK (tag view) https://try.gitea.io/silverwind/remap-css/src/tag/5.0.4:

image

OK (branch view) https://try.gitea.io/silverwind/remap-css/src/branch/master:

image

### Gitea Version

1.19

### Can you reproduce the bug on the Gitea demo site?

Yes

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #21687

Changes in `tag.go`:

`parseTagData` gets called with the input
```
        object a0f5a9a9c6958455c818eecba2199846a96b2612
        type commit
        tag 5.0.4
        tagger silverwind <me@silverwind.io> 1605812507 +0100

        * update deps, fix csstree-validator api (silverwind)
        -----BEGIN PGP SIGNATURE-----

        iQIzBAABCAAdFiEEWPb2jX6FS2mqyJRQLmK0HJOGlEMFAl+2wRsACgkQLmK0HJOG
        lENNjw/+Lyi7Chthc+ZHaFUYJLwZ9a0z/itLkMlQ6XzHIRcsSY0B3KXIIvwta8HV
        k9YcfQyIyWBECOVvUd+sdcPEkKrhT4npPQV31nqUFGONN7jRU9nuY7f4KHZ3mY9L
        KlaRC4ybivfB9qK315NbOVrkG4gckWoGfaTg/0VP4WSbvPVXMKOpRqJC2I+eqiFD
        S4NanKrLuf/pwDcNJ+Ix6QWKYs0lIxJp49QEiVyse0Nj7XwTVRtSc5hz4zrAO72i
        sTHTsiMIkPBCiwo9dzYBA7PjfpPMfJz4BzcUZt0phOU/jGEbDtZ0b16mlBYLl865
        fAuphh3fW3qOSB2eYvDtapLdantnzRg01p25RSThRbG9WDQAphJoIkWW3Evgr1/v
        qISjXNNZFmumz5BVEjfTtT85qdht2nsuwX+Loh7HIDF4Qkd2XpoJFJs5UhK/2B77
        JMZMBq+MwALCrk8Kb6/Kgu959mDtwwB614MHVbxbRLPzqVRLyKisVqA9Fpgj0lXD
        6iRwm9O2tdGskRt+ErfUmiYMwUIQsEwxcBI/iFjwZ756Xs19ZcIvy1e4cN/Mp0Gv
        v037qqQwgFwdIwLwfXGFagcfO7tXARvG5eHhwyzsujY8pZxo8PIKyPBzFqjl0SIE
        sJH6vCQfPoObtsf95OaNVbGaX3YO2uEAhfS4Uq+E558bcfgJzWg=
        =Ko7n
        -----END PGP SIGNATURE-----
```
As you can see, the message is not the same as for the commit. My change builds the expected message.

Changes in other files:
The old code overwrites data of the commit with data from the tag. That's the reason of the broken UI. Because of the wrong output from `parseTagData` the correct values were replaced with wrong values. In my tests both sides of the assignment contained the same values now so I removed the assignments. But I don't know if that's true all the time.

### CODE REVIEW

**Review by silverwind — APPROVED** [review]
Can confirm this works at least for the no-gogit case. Also I fixed the unused `fmt` import.

**Comment by lunny** [comment]
CI failure is related.

**Comment by silverwind** [comment]
I noticed that the same issue is visible on tag messages on a connected Drone instance. I'm not sure if this integration is push or pull, but I certainly wonder if those messages will be fixed there as well.

**Comment by silverwind** [comment]
Checking further in the drone issue, I see that the cut-off message is delivered in `head_commit` for the tag event:

```json
"ref": "refs/tags/v1.9.1",
"head_commit": {
  "id": "24c26a39ac96b6facb6f0ffb7c0cb6826ebef48c",
  "message": "* update js deps"
}
```

Comparing a commit event, it is fine there already:

```json
"ref": "refs/heads/master",
"commits": [
  {
    "id": "24c26a39ac96b6facb6f0ffb7c0cb6826ebef48c",
    "message": "v1.9.1\n\n* update js deps"
  }
"head_commit": {
   "id": "24c26a39ac96b6facb6f0ffb7c0cb6826ebef48c",
   "message": "v1.9.1\n\n* update js deps"
}
```

**Comment by silverwind** [comment]
Can confirm this also fixes the message in webhook tag events. Before:

```
"message": "* test commit (silverwind)",
```

After:
```
"message": "0.2.2\n\n* test commit (silverwind)\n",
```

So I'd say this is a very valuable fix.

**Comment by silverwind** [comment]
Test failure is still related:

```
--- FAIL: TestGetTagCommitWithSignature (0.01s)
    repo_commit_test.go:50: 
          Error Trace:  /drone/src/modules/git/repo_commit_test.go:50
          Error:        Expected value not to be nil.
          Test:         TestGetTagCommitWithSignature
    repo_commit_test.go:52: 
          Error Trace:  /drone/src/modules/git/repo_commit_test.go:52
          Error:        Not equal: 
                        expected: "tag"
                        actual  : "Added short link\n"
                        
                        Diff:
                        --- Expected
                        +++ Actual
                        @@ -1 +1,2 @@
                        -tag
                        +Added short link
                        +
          Test:         TestGetTagCommitWithSignature

**Comment by KN4CK3R** [comment]
I'm not sure anymore this is the solution we want. If there is a signature present, this sig will not match the message now...

**Comment by silverwind** [comment]
Not sure I understand. Isn't the PGP signature supposed to sign the whole commit message, including the first line? At least [this](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work) makes me believe there is no special case that would exclude the first line.

**Comment by KN4CK3R** [comment]
Maybe I'm wrong but from the example in the first post the code now extracts the message
```
5.0.4

* update deps, fix csstree-validator api (silverwind)
```
If we would build the "tag data" again from our object we would end up with something like
```
tag 5.0.4
tagger silverwind <me@silverwind.io> 1605812507 +0100

5.0.4

* update deps, fix csstree-validator api (silverwind)
```
and that would not match the signature. So it may not be desired to modify the message. It may be better to build the new format on the fly with a method like this?
```go
func (t *Tag) String() string {
	if t.Name != "" {
		if t.Message == "" {
			return t.Name
		} else {
			return t.Name + "\n\n" + t.Message
		}
	}
	return t.Message
}
```
I'm a bit puzzled...

**Comment by silverwind** [comment]
Viewing this tag in git cli shows:

1. The commit message has both the version title and the body
2. The tag body itself does not have the version title because it's not part of the tag message

I'm thinking we should not even parse the tag when viewing repo commits as tags are just pointers to commits (with annotated tags also have a message attached).

```
$ git show 5.0.4
tag 5.0.4
Tagger: silverwind <me@silverwind.io>
Date:   Thu Nov 19 20:01:47 2020 +0100

* update deps, fix csstree-validator api (silverwind)
-----BEGIN PGP SIGNATURE-----

iQIzBAABCAAdFiEEWPb2jX6FS2mqyJRQLmK0HJOGlEMFAl+2wRsACgkQLmK0HJOG
lENNjw/+Lyi7Chthc+ZHaFUYJLwZ9a0z/itLkMlQ6XzHIRcsSY0B3KXIIvwta8HV
k9YcfQyIyWBECOVvUd+sdcPEkKrhT4npPQV31nqUFGONN7jRU9nuY7f4KHZ3mY9L
KlaRC4ybivfB9qK315NbOVrkG4gckWoGfaTg/0VP4WSbvPVXMKOpRqJC2I+eqiFD
S4NanKrLuf/pwDcNJ+Ix6QWKYs0lIxJp49QEiVyse0Nj7XwTVRtSc5hz4zrAO72i
sTHTsiMIkPBCiwo9dzYBA7PjfpPMfJz4BzcUZt0phOU/jGEbDtZ0b16mlBYLl865
fAuphh3fW3qOSB2eYvDtapLdantnzRg01p25RSThRbG9WDQAphJoIkWW3Evgr1/v
qISjXNNZFmumz5BVEjfTtT85qdht2nsuwX+Loh7HIDF4Qkd2XpoJFJs5UhK/2B77
JMZMBq+MwALCrk8Kb6/Kgu959mDtwwB614MHVbxbRLPzqVRLyKisVqA9Fpgj0lXD
6iRwm9O2tdGskRt+ErfUmiYMwUIQsEwxcBI/iFjwZ756Xs19ZcIvy1e4cN/Mp0Gv
v037qqQwgFwdIwLwfXGFagcfO7tXARvG5eHhwyzsujY8pZxo8PIKyPBzFqjl0SIE
sJH6vCQfPoObtsf95OaNVbGaX3YO2uEAhfS4Uq+E558bcfgJzWg=
=Ko7n
-----END PGP SIGNATURE-----

commit a0f5a9a9c6958455c818eecba2199846a96b2612 (tag: 5.0.4, local/master, gitea-try/master)
Author: silverwind <me@silverwind.io>
Date:   Thu Nov 19 20:01:47 2020 +0100

    5.0.4

    * update deps, fix csstree-validator api (silverwind)
    * ```

**Comment by silverwind** [comment]
If it helps, the tag messages on my repos are generated [here](https://github.com/silverwind/versions/blob/4a51702af49ef87f9e4adba096f9a4bc64b852c2/versions.js#L363). Note that it supplies the changelog separately as the tag message, and that the tag message excludes the version line present in the commit message. Maybe that is where the confusion comes from. The tag message does not include the first line.

File: modules/git/repo_tag.go [hunk 1]
```diff
@@ -202,6 +202,13 @@ func parseTagRef(ref map[string]string) (tag *Tag, err error) {
 		}
 	}
 
+	tag.Message = strings.TrimSpace(tag.Message)
+	if tag.Message == "" {
+		tag.Message = tag.Name
+	} else if tag.Name != "" {
+		tag.Message = tag.Name + "\n\n" + tag.Message
+	}
```
Comment (silverwind) [inline modules/git/repo_tag.go:210]: I think this joining is incorrect, we should not alter the tag message. What was wrong was just the retrieval of the tagged commit message.

File: modules/git/repo_commit_gogit.go [hunk 2]
```diff
@@ -112,6 +79,13 @@ func (repo *Repository) getCommit(id SHA1) (*Commit, error) {
 			}
 		}
 		if err == nil {
+			tagObject.Message = strings.TrimSpace(tagObject.Message)
+			if tagObject.Message == "" {
+				tagObject.Message = tagObject.Name
+			} else if tagObject.Name != "" {
+				tagObject.Message = tagObject.Name + "\n\n" + tagObject.Message
+			}
```
Comment (silverwind) [inline modules/git/repo_commit_gogit.go:87]: Same here, I think we should not alter the tag message.

File: modules/git/repo_commit_nogogit.go [hunk 3]
```diff
@@ -108,7 +108,7 @@ func (repo *Repository) getCommitFromBatchReader(rd *bufio.Reader, id SHA1) (*Co
 			return nil, err
 		}
 
-		commit.CommitMessage = strings.TrimSpace(tag.Message)
+		commit.CommitMessage = tag.Message
 		commit.Author = tag.Tagger
 		commit.Signature = tag.Signature
```
Comment (silverwind) [inline modules/git/repo_commit_nogogit.go:113]: Why are we overwriting the commit's data here with incorrect data from the tag? I think this is the root of the problem and I guess if we just remove these three lines, the issue may be fixed.

**Comment by silverwind** [comment]
Alternative PR: https://github.com/go-gitea/gitea/pull/21804

**Comment by KN4CK3R** [comment]
I added back the overwriting because of the tests and that was the point where I started to have doubts.

**Comment by silverwind** [comment]
I see my changes are pretty similar to your first PR, minus the changes in `modules/git/tag.go`, we'll see if I run into the same issues, but adding back the overwriting is definitely not the right fix.

**Comment by silverwind** [comment]
The problem you might have been facing is that the test commit for `TestGetTagCommitWithSignature` is not actually signed properly as verified by `git verify-commit`. Only the `test` tag is signed properly. With my new commit, both tag and commit are signed correctly.

**Pull request closed without merging**